MudStepper: Fix final step completed color

Lewis-Pitman opened the pull request:
## Description
When the MudStepper is linear, the final step will now show the completed colour once completed.

Originally, it was impossible for the last step in a linear MudStepper to show the completed step color, as the last step was also the current step. This has been fixed so that if the MudStepper is linear and the last step has been completed, the current step color changes to the completed step color. The current step color is restored if the user resets the progress of the stepper. 

This behaviour doesn't happen in a non linear MudStepper, as these allow you to go back and view other steps when all steps have been completed, and so having a different colour representing the current step is necessary.

Resolves #10962 

## How Has This Been Tested?
No tests have been added.  The code does not alter existing code, and functions as expected on MudBlazor.Docs.Server

## Type of Changes
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (fix or improvement to the website or code docs)

![MudStepper_final_step_complete_color](https://github.com/user-attachments/assets/c0bab47b-c843-449f-b794-254457b8db8e)

## Checklist
- [X] The PR is submitted to the correct branch (`dev`).
- [X] My code follows the code style of this project.
- [ ] I've added relevant tests.

Lewis-Pitman committed:
Fix final step's completed color

Fixed final step's completed color. When the MudStepper is linear, the final step will now show the completed colour once completed.

Anu6is commented on src/MudBlazor/Components/Stepper/MudStepper.razor.cs:
```diff
@@ -467,11 +485,22 @@ private async Task SetActiveIndexAsync(int value, bool skipDisabled = false)
 
+    private async Task SetCurrentStepColorAsync(Color color)
+    {
+        CurrentStepColor = color;
```
Don't set `CurrentStepColor` directly, calling `SetValueAsync` should be enough.

Lewis-Pitman replied:
Hi, just took a quick look during my lunch break. Only using SetValueAsync doesn't seem to be working here. Have I implemented it wrong? 

![image](https://github.com/user-attachments/assets/d2747c5b-bfcc-4d25-a964-eb73c2ecd66e)

Edit: I think I know the solution, will implement when I can

ScarletKuro replied:
> Hi, just took a quick look during my lunch break. Only using SetValueAsync doesn't seem to be working here. Have I implemented it wrong?

All places that used `CurrentStepColor` needs to be replaced with `_currentStepColor.Value` like this place:
https://github.com/MudBlazor/MudBlazor/blob/e5f69083a7c40730de2a4635a9f67e927f834d2f/src/MudBlazor/Components/Stepper/MudStep.cs#L51
if it's parent component then `Parent?.GetState(x => x.CurrentStepColor)`.

Anu6is commented on src/MudBlazor/Components/Stepper/MudStepper.razor.cs:
```diff
@@ -71,6 +80,13 @@ public MudStepper()
+    /// </summary>
+    [Parameter]
+    [Category(CategoryTypes.List.Behavior)]
+    public EventCallback<Color> CurrentStepColorChanged { get; set; }
```
Is this `EventCallback` necessary?

Anu6is reviewed (commented):
There should be a new test to confirm that the final step color is actually updated once completed

Lewis-Pitman commented:
Hi, thanks for the review. This is my first PR so I was mainly trying to follow the existing style. I'll look into implementing what you've said

ScarletKuro commented on src/MudBlazor/Components/Stepper/MudStepper.razor.cs:
```diff
@@ -22,9 +22,18 @@ public MudStepper()
+        _currentStepColor = registerScope.RegisterParameter<Color>(nameof(CurrentStepColor))
+            .WithParameter(() => CurrentStepColor)
+            .WithEventCallback(() => CurrentStepColorChanged)
+            .WithChangeHandler(async args => await SetCurrentStepColorAsync(args.Value));
```
just make it
```
.WithChangeHandler(SetCurrentStepColorAsync);
```
```
private async Task SetCurrentStepColorAsync(ParameterChangedEventArgs<Color> color)
```

Lewis-Pitman committed:
Modify CssBuilder

Modifies CssBuilder instead of changing the current colour parameter

Lewis-Pitman committed:
Fixed when completed color not set

Fixed when completed color isn't set but a different current color is set.

Lewis-Pitman commented:
Sorry, just noticed after pushing that the same bug happens if you don't set a completed color but change the current step color. This new push fixes that

ScarletKuro commented:
> Also, should I add a test for this?

I think in this case a test would be nice

Lewis-Pitman commented:
> > Also, should I add a test for this?
> 
> I think in this case a test would be nice
> rest looks good to me

Great, I'll look at implementing this. 

Should I worry about the failing check? It's happening because the Parent? Null check isn't necessary, but it does follow the existing style

ScarletKuro commented:
> It's happening because the Parent? Null check isn't necessary, but it does follow the existing style

Leave as is

henon reviewed (changes requested):
Great! This can be merged once we have a test case that fails if you undo your fix and passes if you add it back in.

Lewis-Pitman committed:
Add test

Added a test to ensure the last step has it's completed color if it is linear, and does not have the current color if a completed color was not set.

henon commented:
Thanks @Lewis-Pitman
